src/MongoDB.EntityFrameworkCore/Storage/MongoDatabaseWrapper.cs

```diff
@@ -217,4 +148,148 @@ private static string GetId(IUpdateEntry entry)
             _ => throw new NotSupportedException($"Unknown type for _id property on entity type ${entry.EntityType.DisplayName()}")
         };
     }
+
+    private IEnumerable<MongoUpdate> ConvertUpdateEntriesToMongoUpdates(IEnumerable<IUpdateEntry> entries)
+    {
+        return
+            GetAllChangedRootEntries(entries)
+            .Select(ConvertUpdateEntryToMongoUpdate)
+            .Where(x => x != null)
+            .Cast<MongoUpdate>();
+    }
+
+    private MongoUpdate? ConvertUpdateEntryToMongoUpdate(IUpdateEntry entry)
+    {
+        var collectionName = entry.EntityType.GetCollectionName();
+        var id = GetId(entry);
+
+        return entry.EntityState switch
+        {
+            EntityState.Added => ConvertAddedEntryToMongoUpdate(collectionName, id, entry),
+            EntityState.Deleted => ConvertDeletedEntryToMongoUpdate(collectionName, id, entry),
+            EntityState.Detached => null,
+            EntityState.Modified => ConvertModifiedEntryToMongoUpdate(collectionName, id, entry),
+            EntityState.Unchanged => null,
+            _ => throw new NotSupportedException($"Unexpected entity state: {entry.EntityState}.")
+        };
+    }
+
+    private MongoUpdate ConvertAddedEntryToMongoUpdate(string collectionName, string id, IUpdateEntry entry)
+    {
+        throw new NotImplementedException(); // EF-17
+    }
+
+    private MongoUpdate ConvertDeletedEntryToMongoUpdate(string collectionName, string id, IUpdateEntry entry)
+    {
+        var filter = new BsonDocument("_id", id);
+        var model = new DeleteOneModel<BsonDocument>(filter);
+        return new MongoUpdate(collectionName, model);
+    }
+
+    private MongoUpdate ConvertModifiedEntryToMongoUpdate(string collectionName, string id, IUpdateEntry entry)
+    {
+        throw new NotImplementedException(); // EF-15
+    }
+
+    private long SaveMongoUpdates(IEnumerable<MongoUpdate> updates)
+    {
+        var database = _mongoClient.Database;
+        var client = database.Client;
+        using var session = client.StartSession();
+        return session.WithTransaction((session, cancellationToken) => SaveMongoUpdates(session, database, updates));
+    }
+
+    private long SaveMongoUpdates(IClientSessionHandle session, IMongoDatabase database, IEnumerable<MongoUpdate> updates)
+    {
+        var documentsAffected = 0L;
+        foreach (var batch in BatchUpdatesByCollection(updates))
```

Comment:
Since the current MongoDB bulk write API only supports bulk writes to a single collection we have to break up the updates by collection.

I'm assuming order matters and therefore not reordering updates (so different updates to the same collection will end up in different batches if interrupted by updates to other collections).

Reply:
Ordering should not matter in our provider as we don't have FK constraints, cascading deletes etc. that SQL has to worry about.